File: 08_07_Permutation_without_Dups.cpp
/** Solution 1 **/

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> return_array;
        permutation_helper(nums, 0, return_array);
        return return_array;
    }

    void permutation_helper(vector<int> & nums, int current_idx, vector<vector<int>> & result) {
        if (current_idx == nums.size()) {
            result.push_back(nums);
        } else {
            for (auto i = current_idx; i < nums.size(); i++) {
                swap(nums[current_idx], nums[i]);
                permutation_helper(nums, current_idx + 1, result);
                swap(nums[current_idx], nums[i]);                
            }
        }
    }
};

/** Solution 2 **/

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> return_array;
        permutation_helper(vector<int>(), nums, return_array);
        return return_array;
    }

    void permutation_helper(vector<int> current_set, vector<int> nums, vector<vector<int>> & result) {
        if (current_set.size() == nums.size()) {
            result.push_back(current_set);
        } else {
            for (auto i = 0; i < nums.size(); i++) {
                if (find(current_set.begin(), current_set.end(), nums[i]) == current_set.end()) {
                    current_set.push_back(nums[i]);
                    permutation_helper(current_set, nums, result);
                    current_set.pop_back();
                }
            }
        }
    }
};

File: 08_07_Permutation_without_Dups.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> reVal = new ArrayList<>();
        
        permuteHelper(nums, new ArrayList<>(), reVal);
        
        return reVal;
    }
    
    
    void permuteHelper(int[] nums, List<Integer> currentList, List<List<Integer>> reVal) {
        
        if (currentList.size() == nums.length) {
            reVal.add(new ArrayList(currentList));
        }
        
        for (int i = 0;i < nums.length;i++) {
            if (currentList.contains(nums[i])) {
                continue;
            }
            
            currentList.add(nums[i]);
            permuteHelper(nums, currentList, reVal);
            currentList.remove(currentList.size() - 1);
        }
    }
}
